packages/fxa-auth-server/lib/routes/account.js

```diff
@@ -1512,19 +1513,51 @@ module.exports = (
 
         const { uid, scope } = request.auth.credentials;
         const { ecosystemAnonId } = request.payload;
-        const noneMatchHeader = request.headers['If-None-Match'];
         const scopeSet = ScopeSet.fromArray(scope);
 
+        function parseMultiValueHeader(named) {
+          const value = request.headers[named];
+          if (!value || !value.length) {
+            return [];
+          }
+          return value.toString().split(/,\s?/);
+        }
+
+        function hashAnonId(anonId) {
+          const hash = crypto.createHash('md5');
+          return hash.update(anonId).digest('hex');
+        }
+
+        const ifNoneMatch = parseMultiValueHeader('If-None-Match');
+        const ifMatch = parseMultiValueHeader('If-Match');
+
         if (!scopeSet.contains('profile:ecosystem_anon_id:write')) {
           throw error.invalidScopes(scopeSet);
         }
 
         await customs.check(request, uid, 'updateEcosystemAnonId');
 
-        if (noneMatchHeader === '*') {
-          const account = await db.account();
-          if (account.ecosystemAnonId != null) {
-            throw error.anonIdExists();
+        if (ifNoneMatch.length || ifMatch.length) {
+          const { ecosystemAnonId: existingAnonId } = await db.account();
+
+          if (existingAnonId != null && existingAnonId.length) {
+            const hashedAnonId = hashAnonId(existingAnonId);
+
+            if (ifNoneMatch.includes('*')) {
```

Comment:
I'd evaluate if-match headers first, following the precedence rules described in the [HTTP spec](https://tools.ietf.org/html/rfc7232#section-6).